src/sagemaker_xgboost_container/algorithm_mode/callback.py

```diff
@@ -0,0 +1,205 @@
+    etc. However, this could result in a large number of files to save locally
+    (and on S3 if used in SageMaker). To save disk space and reduce the amount
+    of data required to be downloaded on resumption, we retain only the N
+    (default 5, spcified by 'max_to_keep') most recent checkpoints.
+
+    To delete stale checkpoints, we use a producer-consumer pattern: we start a
+    daemon thread in the background and maintain a queue of files to delete.
+    There may be a lock on the file if SageMaker is uploading the file; in that
+    case, the file is put back on the queue and we try again later. When
+    training is complete, we put SENTINEL on the queue, and when we see the
+    SENTINEL, we clean up and exit the thread.
+    """
+
+    SENTINEL = None
+
+    def __init__(
+            self, checkpoint_dir: str, max_to_keep: int = 5, start_iteration: int = 0,
+            num_round: Optional[int] = None
+            ) -> None:
+        """
+        Parameters
+        ----------
+        checkpoint_dir: indicates the path to the directory where checkpoints
+            will be saved.  Defaults to /opt/ml/checkpoints on SageMaker.
+        max_to_keep: indicates the maximum number of recent checkpoint files to
+            keep.  As new files are created, older files are deleted.  Defaults
+            to 5 (that is, the 5 most recent checkpoint files are kept.)
+        start_iteration: indicates the round at which the current training
+            started. If xgb_model was loaded from a previous checkpoint, this
+            will be greater than 0 (that is, if the previous training ended
+            after round 19, start_iteration will be 20).
+        num_round: (optional) indicates the number of boosting rounds.
+        """
+        self.checkpoint_dir = checkpoint_dir
+        self.max_to_keep = max_to_keep
+        self.start_iteration = start_iteration
+        self.num_round = num_round
+
+        self.step = start_iteration
+        self.previous_checkpoints = [
+            os.path.join(self.checkpoint_dir, f)
+            for f in os.listdir(self.checkpoint_dir)]
+
+        self.thread = None
+        self.delete_queue = queue.Queue()
+
+        self.start()
+
+    def __call__(self, env: CallbackEnv) -> Callable[[CallbackEnv], None]:
+
+        if not os.path.exists(self.checkpoint_dir):
+            os.makedirs(self.checkpoint_dir)
+
+        return self.callback(env)
+
+    def fmt_path(self, i: int) -> str:
+
+        filename = "{}.{}".format(CHECKPOINT_FILENAME, _zero_pad(i))
+        checkpoint_path = os.path.join(self.checkpoint_dir, filename)
+
+        return checkpoint_path
+
+    def start(self) -> None:
+
+        def _delete_once(skip_locked_files=True):
+
+            for i in iter(self.delete_queue.get, self.SENTINEL):
+
+                path = self.fmt_path(i)
+                if (skip_locked_files
+                        and os.path.isfile(path + FILE_LOCK_SUFFIX)
```

Comment:
Can you move these two ```.isfile()``` checks to a well named method for readability? I would also re-document the behavior here; whenever we have asynchronous processes I'd like to make sure its really clear what's happening.